carbon-design-system/carbon-addons-iot-react
# 2603 column customization

Closes #2603
[SEE PR UPDATE HERE](#PR-FEEDBACK-UPDATE)

**Summary**
Adds a new modal for column customization. See specs listed in #2603. In addition to those specs I've also added support to hide/show selected columns and to handle existing column grouping. 

**Change List (commits, features, bugs, etc)**
Except for the changes made in separate PRs the following notable changes were needed to existing components:

- [list-item.scss] Correctly position the item actions in RTL. This also solves issue #3069
- [List.story] Update the overflow flipped in RTL since it was needed once the actions where correctly positioned
- [HierarchyList] Change to only render the HierarchyListReorderModal if there are any items selected. Makes more sense, is more efficient and makes testing easier.
- [ListBuilder] Added a bunch of props that needed to be exposed to the TableColumnCustomizationModal. The most notable is the `useCheckboxes` which turns the left list into List component with checkboxes. The default appearance of the ListBuilder is left intact, but it can via props be made to match the new visual design.

**Acceptance Test (how to verify the PR)**
_Documentation_
Please check spelling, general quality and coverage of the added documentation

_ListBuilder_
1. Go to https://deploy-preview-3073--carbon-addons-iot-react.netlify.app?path=/story/2-watson-iot-experimental-☢%EF%B8%8F-listbuilder--stateful-example-with-checkboxes
2. Verify that it has the new design and works as expected
3. NOTE, clicking on a checkbox will add the columns twice until [#3066 ? with fix](https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3066) is merged

_Table integration_
1. Go to https://deploy-preview-3073--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-table-table--with-column-customization-modal
2. Play around with the example and the knobs. And make sure the behaviour matches the specification. 
3. There is no written spec for column grouping and this modal, but make sure that columns can't be dropped outside their group and that groups can't be nested. Hiding a group hides all the columns and hiding all the elements will also show the group as hidden.
4. Make sure the "Load more" function is working
5. StatefulTable should work the same as Table

_TableColumnCustomizationModal stories_
1. Go to https://deploy-preview-3073--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table-☢%EF%B8%8F-tablecolumncustomizationmodal--experimental
2. Go through all the stories and knobs and verify that they work as expected and makes sense

**Regression Test (how to make sure this PR doesn't break old functionality)**
- Unit tests and e2e not failing
- Old functionality in List, HierarchyList and ListBuilder still works as before



<a name="PR-FEEDBACK-UPDATE"></a>
**PR FEEDBACK UPDATE**

Row height and alignment etc of tall rows and other list related styles will be fixed in #3062 so please ignore for now.

1. Made the primary and secondary item value configurable to be either column `name` or `id`. The secondary value prop can also be undefined. See https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/commits/0a3749ceedcb0f40eb955683596dcd4be0d127e0
2. Search will now include both primary and, if used in the list, secondary value. See https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/commits/86dcd32ddd266da186a74a8aaa748578e09ac99c#diff-658475b36715abfa40bb32ffe205c6e6d3a7a2e4f2eebd9bed9e752b1b5840f9R309-R326
3. Created an async wrapper component to allow the availableColumns to be a promise and handle loading states. See https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/commits/93f077115b5a2fb7a7545e926abae98e9fc83f1b 
4. Show groups in the Available columns list. See https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/commits/86dcd32ddd266da186a74a8aaa748578e09ac99c
5. Add back add deprecate props in ListBuilder. See https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/commits/29397a7720d50427792a3b1ab0aefd35ef0b3be2
6. Update the docs. This is better viewed in its whole, see https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3073/files?authenticity_token=%2FktiFAd97oZ8Ct5le9mPwc4FpN%2BYilWfMXAZhux1GJdUqNNQ9fiD8mh%2Bxei%2FMvBNYetHIYjvsrWTILJzeIap%2Bw%3D%3D&file-filters%5B%5D=.mdx

*Additional test instructions*
1. Test props primaryValue and secondaryValue in https://deploy-preview-3073--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table-%E2%98%A2%EF%B8%8F-tablecolumncustomizationmodal--playground-with-knobs
2. Test updated search using secondaryValue in https://deploy-preview-3073--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table-%E2%98%A2%EF%B8%8F-tablecolumncustomizationmodal--with-many-columns-and-load-more
3. Test knobs and functionality of new async story https://deploy-preview-3073--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table-%E2%98%A2%EF%B8%8F-tablecolumncustomizationmodal--async-with-loading-state
4. For all column customisation stories that can show column grouping. Make sure the groups are properly shown in the "Available columns" list and that the checkboxes of the group have the correct states depending on their children (checked, unchecked, indeterminate). Make sure that selecting a group also selects all children.

**Not done**
- The "pixel perfect" changes needed to the subcomponents (ListBuilder, ComposedModal, List etc) will be fixed in #3062. For now the general style of the underlying components have been kept.
- The checkboxes are firing twice, will be fixed by #3100
- There are some Accessibility issues inherited from the HierarchyList that will be addressed in separate issue.



**Things to look for during review**
